stan/math/prim/functor/apply_scalar_ternary.hpp

```diff
@@ -0,0 +1,216 @@
+
+#include <stan/math/prim/fun/as_column_vector_or_scalar.hpp>
+#include <stan/math/prim/meta/is_stan_scalar.hpp>
+#include <stan/math/prim/meta/is_container.hpp>
+#include <stan/math/prim/meta/is_eigen.hpp>
+#include <stan/math/prim/meta/require_generics.hpp>
+#include <stan/math/prim/err/check_matching_dims.hpp>
+#include <stan/math/prim/err/check_matching_sizes.hpp>
+#include <stan/math/prim/fun/num_elements.hpp>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+/**
+ * Base template function for vectorization of ternary scalar functions
+ * defined by applying a functor to a combination of scalars,
+ * containers of matching sizes, or a combination of a scalar and a container.
+ * These containers can be a standard library vector, Eigen dense
+ * matrix expression template, or container of these. For each specialisation,
+ * the same type as the largest (dimension) input is returned.
+ *
+ * This base template function takes (and returns) scalars.
+ *
+ * @tparam T1 Type of first scalar to which functor is applied.
+ * @tparam T2 Type of second scalar to which functor is applied.
+ * @tparam T3 Type of third scalar to which functor is applied.
+ * @tparam F Type of functor to apply.
+ * @param x First input to which operation is applied.
+ * @param y Second input to which operation is applied.
+ * @param z Third scalar to which operation is applied.
+ * @param f functor to apply to inputs.
+ * @return Scalar with result of applying functor to inputs.
+ */
+template <typename T1, typename T2, typename T3, typename F,
+          require_all_stan_scalar_t<T1, T2, T3>* = nullptr>
+inline auto apply_scalar_ternary(const T1& x, const T2& y, const T3& z,
+                                 const F& f) {
+  return f(x, y, z);
+}
+
+/**
+ * Specialisation for use with three Eigen inputs. Eigen's ternaryExpr framework
+ * is used for more efficient indexing of both row- and column-major inputs
+ * without separate loops.
+ *
+ * @tparam T1 Type of first argument to which functor is applied.
+ * @tparam T2 Type of second argument to which functor is applied.
+ * @tparam T3 Type of third argument to which functor is applied.
+ * @tparam F Type of functor to apply.
+ * @param x First Eigen input to which operation is applied.
+ * @param y Second Eigen input to which operation is applied.
+ * @param z Third Eigen input to which operation is applied.
+ * @param f functor to apply to Eigen input.
+ * @return Eigen object with result of applying functor to inputs.
+ */
+template <typename T1, typename T2, typename T3, typename F,
+          require_all_eigen_t<T1, T2, T3>* = nullptr>
+inline auto apply_scalar_ternary(const T1& x, const T2& y, const T3& z,
+                                 const F& f) {
+  check_matching_dims("Ternary function", "x", x, "y", y);
+  check_matching_dims("Ternary function", "y", y, "z", z);
+  check_matching_dims("Ternary function", "x", x, "z", z);
+  return x.ternaryExpr(y, z, f);
```

Comment:
I think we want `make_holder` here

```suggestion
  return make_holder(
      [](auto& f_inner, const auto& x_inner, const auto& y_inner, const auto& z_inner) {
        return x_inner.ternaryExpr(f_inner, y_inner, z_inner);
      },
      std::forward<F>(f), std::forward<T1>(x), std::forward<T2>(y), std::forward<T3>(z));
```